File: src/services/ml/pipeline.ts
/**
 * ML Pipeline Orchestrator
 * Coordinates the entire ML workflow for resume processing and matching
 */

import { logger } from '@/utils/logger';
import { extractTextFromFile, parseStructuredData } from './textExtraction';
import { extractSkills } from './skillExtraction';
import { generateEmbedding } from './embeddings';
import { matchResumeToJobs, matchJobToResumes, type MatchResult } from './matchingEngine';
import type { Resume, Job } from '@/types/models';
import { generateId } from '@/utils/helpers';

export interface ResumeProcessingResult {
  resumeId: string;
  extractedText: string;
  skills: string[];
  education: any[];
  experience: any[];
  embedding: number[];
  status: 'completed' | 'failed';
  error?: string;
}

export interface JobProcessingResult {
  jobId: string;
  skills: string[];
  embedding: number[];
  status: 'completed' | 'failed';
  error?: string;
}

/**
 * Process uploaded resume through the entire ML pipeline
 */
export async function processResume(file: File, userId: string): Promise<ResumeProcessingResult> {
  const resumeId = generateId();
  
  logger.info('Starting resume processing pipeline', {
    resumeId,
    userId,
    fileName: file.name,
  });

  try {
    // Step 1: Extract text from file
    logger.time(`resume-extraction-${resumeId}`);
    const { text: extractedText } = await extractTextFromFile(file);
    logger.timeEnd(`resume-extraction-${resumeId}`);

    // Step 2: Parse structured data
    logger.time(`resume-parsing-${resumeId}`);
    const structured = parseStructuredData(extractedText);
    logger.timeEnd(`resume-parsing-${resumeId}`);

    // Step 3: Extract skills
    logger.time(`skill-extraction-${resumeId}`);
    const { skills } = extractSkills(extractedText);
    logger.timeEnd(`skill-extraction-${resumeId}`);

    // Step 4: Generate embedding
    logger.time(`embedding-generation-${resumeId}`);
    const embedding = await generateEmbedding(extractedText);
    logger.timeEnd(`embedding-generation-${resumeId}`);

    logger.info('Resume processing completed', {
      resumeId,
      skillCount: skills.length,
      textLength: extractedText.length,
    });

    return {
      resumeId,
      extractedText,
      skills,
      education: structured.education,
      experience: structured.experience,
      embedding,
      status: 'completed',
    };
  } catch (error) {
    logger.error('Resume processing failed', error as Error, { resumeId });
    
    return {
      resumeId,
      extractedText: '',
      skills: [],
      education: [],
      experience: [],
      embedding: [],
      status: 'failed',
      error: (error as Error).message,
    };
  }
}

/**
 * Process job description through the ML pipeline
 */
export async function processJobDescription(jobData: {
  title: string;
  description: string;
  requiredSkills: string[];
}): Promise<JobProcessingResult> {
  const jobId = generateId();
  
  logger.info('Starting job processing pipeline', {
    jobId,
    title: jobData.title,
  });

  try {
    // Combine job data for embedding
    const combinedText = `${jobData.title}\n${jobData.description}\nRequired Skills: ${jobData.requiredSkills.join(', ')}`;

    // Extract additional skills from description
    const { skills: extractedSkills } = extractSkills(combinedText);
    
    // Combine with provided skills
    const allSkills = Array.from(new Set([...jobData.requiredSkills, ...extractedSkills]));

    // Generate embedding
    const embedding = await generateEmbedding(combinedText);

    logger.info('Job processing completed', {
      jobId,
      skillCount: allSkills.length,
    });

    return {
      jobId,
      skills: allSkills,
      embedding,
      status: 'completed',
    };
  } catch (error) {
    logger.error('Job processing failed', error as Error, { jobId });
    
    return {
      jobId,
      skills: [],
      embedding: [],
      status: 'failed',
      error: (error as Error).message,
    };
  }
}

/**
 * Execute matching workflow for a resume against all jobs
 */
export async function executeResumeMatching(
  resume: Resume,
  allJobs: Job[],
  topK: number = 10
): Promise<MatchResult[]> {
  logger.info('Executing resume matching workflow', {
    resumeId: resume.id,
    jobCount: allJobs.length,
    topK,
  });

  try {
    // Filter active jobs only
    const activeJobs = allJobs.filter(job => job.status === 'active');

    if (activeJobs.length === 0) {
      logger.warn('No active jobs available for matching');
      return [];
    }

    // Execute matching
    const matches = matchResumeToJobs(resume, activeJobs, topK);

    logger.info('Resume matching completed', {
      resumeId: resume.id,
      matchCount: matches.length,
    });

    return matches;
  } catch (error) {
    logger.error('Resume matching failed', error as Error, {
      resumeId: resume.id,
    });
    throw error;
  }
}

/**
 * Execute matching workflow for a job against all resumes
 */
export async function executeJobMatching(
  job: Job,
  allResumes: Resume[],
  topK: number = 50
): Promise<MatchResult[]> {
  logger.info('Executing job matching workflow', {
    jobId: job.id,
    resumeCount: allResumes.length,
    topK,
  });

  try {
    // Filter completed resumes only
    const completedResumes = allResumes.filter(resume => resume.status === 'completed');

    if (completedResumes.length === 0) {
      logger.warn('No completed resumes available for matching');
      return [];
    }

    // Execute matching
    const matches = matchJobToResumes(job, completedResumes, topK);

    logger.info('Job matching completed', {
      jobId: job.id,
      matchCount: matches.length,
    });

    return matches;
  } catch (error) {
    logger.error('Job matching failed', error as Error, {
      jobId: job.id,
    });
    throw error;
  }
}

/**
 * Health check for ML pipeline
 */
export async function mlPipelineHealthCheck(): Promise<{
  status: 'healthy' | 'degraded' | 'unhealthy';
  components: {
    textExtraction: boolean;
    skillExtraction: boolean;
    embedding: boolean;
    matching: boolean;
  };
}> {
  logger.info('Running ML pipeline health check');

  const components = {
    textExtraction: true,
    skillExtraction: true,
    embedding: true,
    matching: true,
  };

  try {
    // Test skill extraction
    const testText = 'Python JavaScript React Node.js';
    const { skills } = extractSkills(testText);
    components.skillExtraction = skills.length > 0;

    // Test embedding generation
    const testEmbedding = await generateEmbedding(testText);
    components.embedding = testEmbedding.length > 0;

    const healthyCount = Object.values(components).filter(Boolean).length;
    const status = healthyCount === 4 ? 'healthy' : healthyCount >= 2 ? 'degraded' : 'unhealthy';

    logger.info('ML pipeline health check completed', { status, components });

    return { status, components };
  } catch (error) {
    logger.error('ML pipeline health check failed', error as Error);
    return { status: 'unhealthy', components };
  }
}

File: backend/app/pipeline/pipeline.py
"""
ML Pipeline orchestrator for processing resumes and jobs.
"""
import uuid
from datetime import datetime
from typing import Dict, Any, Optional
import structlog

from .text_extraction import extract_text_from_file, extract_text_from_job_description
from .skill_extraction import extract_skills_from_resume, extract_skills_from_job
from .embeddings import generate_embedding

logger = structlog.get_logger(__name__)


class ProcessingResult:
    """Result of processing a document."""

    def __init__(
        self,
        status: str = "pending",
        resume_id: Optional[str] = None,
        job_id: Optional[str] = None,
        extracted_text: Optional[str] = None,
        skills: Optional[list] = None,
        education: Optional[list] = None,
        experience: Optional[list] = None,
        embedding: Optional[list] = None,
        error: Optional[str] = None,
    ):
        self.status = status
        self.resume_id = resume_id
        self.job_id = job_id
        self.extracted_text = extracted_text
        self.skills = skills or []
        self.education = education or []
        self.experience = experience or []
        self.embedding = embedding or []
        self.error = error


async def process_resume(
    file_content: bytes,
    filename: str,
    user_id: str
) -> ProcessingResult:
    """
    Process a resume file through the complete ML pipeline.

    Args:
        file_content: Raw file bytes
        filename: Original filename
        user_id: User ID who uploaded the resume

    Returns:
        ProcessingResult with extracted data
    """
    logger.info("Starting resume processing", filename=filename, user_id=user_id)

    result = ProcessingResult()
    result.resume_id = str(uuid.uuid4())

    try:
        # Step 1: Extract text from file
        logger.info("Extracting text from resume")
        extracted_text = await extract_text_from_file(file_content, filename)

        if not extracted_text:
            result.status = "failed"
            result.error = "No text could be extracted from the file"
            return result

        result.extracted_text = extracted_text

        # Step 2: Extract skills
        logger.info("Extracting skills from resume")
        skills = await extract_skills_from_resume(extracted_text)
        result.skills = skills

        # Step 3: Generate embedding
        logger.info("Generating embedding for resume")
        embedding = await generate_embedding(extracted_text)
        result.embedding = embedding

        # Step 4: Extract additional metadata (simplified)
        # In a full implementation, this would use more sophisticated NLP
        result.education = []  # Would extract education info
        result.experience = []  # Would extract experience info

        result.status = "completed"
        logger.info("Resume processing completed successfully", resume_id=result.resume_id)

    except Exception as e:
        logger.error("Resume processing failed", error=str(e), filename=filename)
        result.status = "failed"
        result.error = str(e)

    return result


async def process_job_description(job_data: Dict[str, Any]) -> ProcessingResult:
    """
    Process a job description through the ML pipeline.

    Args:
        job_data: Job data including title, description, skills

    Returns:
        ProcessingResult with extracted data
    """
    logger.info("Starting job processing", title=job_data.get("title"))

    result = ProcessingResult()
    result.job_id = str(uuid.uuid4())

    try:
        # Combine title and description for processing
        full_text = f"{job_data.get('title', '')} {job_data.get('description', '')}"

        # Step 1: Clean text
        logger.info("Cleaning job description text")
        cleaned_text = await extract_text_from_job_description(full_text)
        result.extracted_text = cleaned_text

        # Step 2: Extract skills (use provided skills or extract from text)
        provided_skills = job_data.get("requiredSkills", [])
        if provided_skills:
            result.skills = provided_skills
        else:
            logger.info("Extracting skills from job description")
            skills = await extract_skills_from_job(cleaned_text)
            result.skills = skills

        # Step 3: Generate embedding
        logger.info("Generating embedding for job")
        embedding = await generate_embedding(cleaned_text)
        result.embedding = embedding

        result.status = "completed"
        logger.info("Job processing completed successfully", job_id=result.job_id)

    except Exception as e:
        logger.error("Job processing failed", error=str(e), title=job_data.get("title"))
        result.status = "failed"
        result.error = str(e)

    return result


async def execute_resume_matching(
    resume_data: Dict[str, Any],
    jobs_data: list
) -> list:
    """
    Execute matching between a resume and multiple jobs.

    Args:
        resume_data: Resume data with skills, embedding, etc.
        jobs_data: List of job data

    Returns:
        List of match results
    """
    from .matching_engine import MatchingEngine

    logger.info(
        "Executing resume-job matching",
        resume_id=resume_data.get("id"),
        job_count=len(jobs_data)
    )

    return await MatchingEngine.batch_calculate_matches([resume_data], jobs_data)


async def execute_job_matching(
    job_data: Dict[str, Any],
    resumes_data: list
) -> list:
    """
    Execute matching between a job and multiple resumes.

    Args:
        job_data: Job data with skills, embedding, etc.
        resumes_data: List of resume data

    Returns:
        List of match results
    """
    from .matching_engine import MatchingEngine

    logger.info(
        "Executing job-resume matching",
        job_id=job_data.get("id"),
        resume_count=len(resumes_data)
    )

    return await MatchingEngine.batch_calculate_matches(resumes_data, [job_data])
